sphinxext/icon.py

```diff
@@ -1,121 +1,337 @@
+
+
+class IconsMMSOrg:
+    """MMS Org icon font"""
+    NODE_NAME = 'IconMMSOrg'
+    ROLE_NAMES = ('icon-mms-org',)
+    CSS_PREFIX = 'mms-org-icon mms-org-icon-'
+    ICONS = set([
+        'activity-feed'
+    ])
 
 
 def class_to_label(css_class):
-    """Approximate a label for screen readers from a FontAwesome class."""
+    """Approximate a label for screen readers from a CSS class name."""
     return PAT_STRIP.sub('', css_class).replace('-', ' ') + ' icon'
 
 
-class IconNode(nodes.General, nodes.Inline, nodes.Element):
-    @staticmethod
-    def visit_icon(self, node):
-        # I've tested this in Safari w/ macOS Voiceover
-        self.body.append(
-            self.starttag(
-                node,
-                'span',
-                CLASS='fas fa-{}'.format(node['css_class']),
-                **{'title': node['label']}))
-        self.body.append(self.starttag(node, 'span', CLASS='screenreader'))
-        self.body.append(node['label'])
+def create_icon_set(app, icons_type):
+    """Create an icon role in the given Sphinx app using the information
+       in the provided class."""
+    class IconNode(nodes.General, nodes.Inline, nodes.Element):
+        """Docutils node for representing an icon."""
+        @staticmethod
+        def visit_icon(self, node):
+            icon_name = node['icon_name']
+            mms_emblem = node['mms_emblem']
+            have_emblem_class = ' {}-{} mms-have-emblem'.format(
+                icons_type.CSS_PREFIX + icon_name,
+                mms_emblem) if mms_emblem else ''
 
-    @staticmethod
-    def depart_icon(self, node=None):
-        self.body.append('</span></span>\n')
+            # I've tested the screenreader behavior in Safari w/ macOS Voiceover
+            self.body.append(
+                self.starttag(
+                    node,
+                    'span',
+                    CLASS=icons_type.CSS_PREFIX + icon_name + have_emblem_class,
+                    **{'title': node['label']}))
+            if mms_emblem:
+                self.body.append(
+                    self.starttag(
+                        node,
+                        'span',
+                        CLASS='mms-emblem mms-emblem-{}'.format(mms_emblem)))
+                self.body.append('</span>')
+            self.body.append(self.starttag(node, 'span', CLASS='screenreader'))
+            self.body.append(node['label'])
 
+        @staticmethod
+        def depart_icon(self, node=None):
+            self.body.append('</span></span>\n')
 
-def icon_role(typ, rawtext, text, lineno, inliner, options={}, content=[]):
-    errors = []
-    has_explicit_title, label, css_class = split_explicit_title(text)
-    if css_class not in ICONS:
-        errors.append(inliner.reporter.error('Unknown icon: ' + css_class, line=lineno))
+    IconNode.__qualname__ = IconNode.__name__ = icons_type.NODE_NAME
+    # Set a global to let Sphinx pickle IconNode by name
+    globals()[icons_type.NODE_NAME] = IconNode
```

Comment:
This is some spooky magic 😅 Might be worthwhile to add a bit more context on why we need to mess with globals?

Reply:
i.e. why do we need to pickle by name?

Reply:
Good thought!